### PR TITLE
Extend the validation function

### PR DESCRIPTION
**Pull request by andrey-churkin**
### Changes

- The validation function can return metrics/logits for each data item from a dataset now.

### Reason for changes

- Improve execution time for algorithm

### Related tickets

N/A

### Tests

- tests/common/accuracy_control/test_evaluator.py

### CODE REVIEW

File: nncf/quantization/algorithms/accuracy_control/evaluator.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+            - values_for_each_item: If the `Evaluator.is_metric_mode()` condition is true,
+                then `values_for_each_item` represents the list of metric value for each item.
+                Otherwise, if the condition is false, it represents list of logits for each
+                item.
+        """
+        model_for_inference = self._algo_backend.prepare_for_inference(model)
+        if self._metric_mode is None:
+            self._determine_mode(model_for_inference, dataset)
+
+        if not self.is_metric_mode() and indices is not None:
```
Comment (alexsu52) [inline nncf/quantization/algorithms/accuracy_control/evaluator.py:90]: Could you briefly explain why `validate` method does not support indices for is_metric_mode == False?

File: nncf/quantization/algorithms/accuracy_control/ranker.py [hunk 2]
```diff
@@ -70,36 +70,42 @@ class GroupToRank:
...
         # We don't need to re-calculate values for the initial model
         # because they don't change. So use this attribute to store
         # them to improve execution time.
-        self._ref_values = None
+        self._reference_values_for_each_item = None
+
+        # A `ranking_fn()` is a function that compares values returned
+        # by `Evaluator.collect_values_for_each_item()` method for
+        # initial and quantized model.
+        if self._evaluator.is_metric_mode():
```
Comment (alexsu52) [inline nncf/quantization/algorithms/accuracy_control/ranker.py:102]: How will this approach be extended if somebody want to implement new ranking function? Why did you refuse from passing `ranking_fn` to `__init__`?

**Review by alexsu52 — COMMENTED** [review]
Please, provide test results on the extended model scope.

File: nncf/quantization/algorithms/accuracy_control/evaluator.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+from typing import Any, Callable, Iterable, List, Optional, Tuple, TypeVar, Union
+
+from nncf.common.factory import EngineFactory
+from nncf.data.dataset import Dataset
+from nncf.quantization.algorithms.accuracy_control.backend import AccuracyControlAlgoBackend
+
+TModel = TypeVar("TModel")
+
+
+class Output:
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:21]: We decided to remove the `Output` class after an offline discussion. The return value from `validation_fn()` will have the following type
```python
TTensor = TypeVar("TTensor")
Tuple[float, Union[None, List[float], List[TTensor], List[List[TTensor]]]]
```

[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:21]: The `TTensor` will be considered as float value if the following type cast is possible `float(tensor)`. I completed the following experiments
```python
# NumPy

x = np.array(1.0, dtype=np.float32)
print(f'Shape: {x.shape}')  # Shape: ()
print(f'float: {float(x)}')  # float: 1.0

x = np.array([1.0], dtype=np.float32)
print(f'Shape: {x.shape}')  # Shape: (1,)
print(f'float: {float(x)}')  # float: 1.0

# PyTorch

x = torch.tensor(1.0, dtype=torch.float32)
print(f'Shape: {x.shape}')  # Shape: torch.Size([])
print(f'float: {float(x)}')  # float: 1.0

x = torch.tensor([1.0], dtype=torch.float32)
print(f'Shape: {x.shape}')  # Shape: torch.Size([1])
print(f'float: {float(x)}')  # float: 1.0

# TensorFlow

x = tf.constant(1.0, dtype=tf.float32)
print(f'Shape: {x.shape}')  # Shape: ()
print(f'float: {float(x)}')  # float: 1.0

x = tf.constant([1.0], dtype=tf.float32)
print(f'Shape: {x.shape}')  # Shape: (1,)
print(f'float: {float(x)}')  # float: 1.0
```

File: nncf/quantization/algorithms/accuracy_control/ranker.py (hunk 2 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/ranker.py]: Fixed

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 1 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:73]: We can not validate (i.e. validation_fn only works for the full dataset) the model on the subset when `is_metric_mode == False`.

File: nncf/openvino/quantization/quantize_model.py [hunk 4]
```diff
@@ -139,7 +137,7 @@ def native_quantize_with_accuracy_control_impl(
     model: ov.Model,
     calibration_dataset: Dataset,
     validation_dataset: Dataset,
-    validation_fn: Callable[[Any, Iterable[Any]], float],
+    validation_fn: Callable[[Any, Iterable[Any]], Tuple[float, Union[List[float], List[Output], None]]],
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/openvino/quantization/quantize_model.py]: Done

File: nncf/openvino/quantization/quantize_model.py [hunk 5]
```diff
@@ -139,7 +138,7 @@ def native_quantize_with_accuracy_control_impl(
     model: ov.Model,
     calibration_dataset: Dataset,
     validation_dataset: Dataset,
-    validation_fn: Callable[[Any, Iterable[Any]], float],
+    validation_fn: Callable[[Any, Iterable[Any]], Tuple[float, Union[None, List[float], List[List[TTensor]]]]],
```
Comment (alexsu52) [inline nncf/openvino/quantization/quantize_model.py:141]: It looks like an adaptation of the POT implementation is required.

File: nncf/quantization/algorithms/accuracy_control/evaluator.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        # +--------------+----------------------+-------------+
+        # | None         | None                 | False       |
+        # +--------------+----------------------+-------------+
+        # | None         | List[float]          | UNEXPECTED  |
+        # +--------------+----------------------+-------------+
+        # | None         | List[List[TTensor]]  | False       |
+        # +--------------+----------------------+-------------+
+
+        self._metric_mode = False
+        if isinstance(metric_value, float) and (
```
Comment (alexsu52) [inline nncf/quantization/algorithms/accuracy_control/evaluator.py:116]: We talked about metric_value should cast to float. I have checked the isinstance with numpy array and it does not work.

File: nncf/quantization/algorithms/accuracy_control/rank_functions.py [hunk 7]
```diff
@@ -9,26 +9,28 @@
...
-from typing import Dict
+from typing import List, TypeVar
 
 import numpy as np
 
+TTensor = TypeVar("TTensor")
 
-def normalized_mse(x_ref: Dict[str, np.ndarray], x_approx: Dict[str, np.ndarray]) -> float:
+
+def normalized_mse(ref_outputs: List[TTensor], approx_outputs: List[TTensor]) -> float:
```
Comment (alexsu52) [inline nncf/quantization/algorithms/accuracy_control/rank_functions.py:19]: Why use `TTensor` here if your code below use backend specific Tensor methods?

**Comment by alexsu52** [comment]
Please, add tests for `validation_fn` as well when finction signature will be approved.

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 6 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py]: Thanks! I fixed it.

File: nncf/quantization/algorithms/accuracy_control/rank_functions.py (hunk 7 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/rank_functions.py]: Thanks! I have improved these parts.

**Comment by andrey-churkin** [comment]
> Please, provide test results on the extended model scope.

Build: #9600

File: nncf/openvino/quantization/quantize_model.py (hunk 5 above)
Comment (andrey-churkin) [reply nncf/openvino/quantization/quantize_model.py:141]: Done

File: nncf/openvino/pot/engine.py [hunk 8]
```diff
@@ -100,7 +100,7 @@ def calc_per_sample_metrics(
 ) -> List[Dict[str, Any]]:
     per_sample_metrics = []
     for inputs in dataset.get_data(subset_indices):
-        value = val_func(compiled_model, [inputs])
+        value, _ = val_func(compiled_model, [inputs])
```
Comment (alexsu52) [inline nncf/openvino/pot/engine.py:103]: Could you share the test plan for changes related to wrapper for POT quantization with accuracy control?

Comment (andrey-churkin) [reply nncf/openvino/pot/engine.py:103]: Build # 9949

**Comment by andrey-churkin** [comment]
run pytorch pre-commit tests

File: nncf/quantization/algorithms/accuracy_control/evaluator.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        # | None         | List[float]          | UNEXPECTED  |
+        # +--------------+----------------------+-------------+
+        # | None         | List[List[TTensor]]  | False       |
+        # +--------------+----------------------+-------------+
+
+        self._metric_mode = False
+        if isinstance(metric_value, float) and (values_for_each_item is None or convert_to_float_possible):
+            self._metric_mode = True
+        elif values_for_each_item is not None and not isinstance(values_for_each_item[0], list):
+            raise RuntimeError("Unexpected return value from provided validation function.")
```
Comment (KodiaqQ) [inline nncf/quantization/algorithms/accuracy_control/evaluator.py:145]: From the user's perspective, this message is not clear. What should they do in this case?

File: nncf/quantization/algorithms/accuracy_control/evaluator.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+    def collect_values_for_each_item(
+        self, model: Any, dataset: Dataset, indices: Optional[List[int]] = None
```
Comment (KodiaqQ) [inline nncf/quantization/algorithms/accuracy_control/evaluator.py:148]:
```suggestion
        self, model: TModel, dataset: Dataset, indices: Optional[List[int]] = None
```

File: nncf/quantization/algorithms/accuracy_control/rank_functions.py [hunk 11]
```diff
@@ -9,26 +9,43 @@
...
+
+
+def create_normalized_mse_func(backend: BackendType) -> Callable[[List[TTensor], List[TTensor]], float]:
+    """
+    Factory method to create backend-specific implementation of the normalized_nmse.
+
+    :param backend: A backend type.
+    :return: The backend-specific implementation of the normalized_nmse.
+    """
+    if backend in (BackendType.ONNX, BackendType.OPENVINO):
```
Comment (KodiaqQ) [inline nncf/quantization/algorithms/accuracy_control/rank_functions.py:28]: Do we have accuracy control method for ONNX?

File: nncf/quantization/algorithms/accuracy_control/ranker.py [hunk 12]
```diff
@@ -154,24 +161,36 @@ def rank_groups_of_quantizers(
...
+            nncf_logger.info("Collecting metrics for each data item using a quantized model")
+            approximate_values_for_each_item = self._evaluator.collect_values_for_each_item(
+                quantized_model, self._dataset
+            )
 
         # Create a subset of data items that will be used to rank groups of quantizers.
-        scores = [self._ranking_fn(ref_val, approx_val) for ref_val, approx_val in zip(self._ref_values, approx_values)]
+        scores = [
+            self._ranking_fn(ref_val, approx_val)
+            for ref_val, approx_val in zip(self._reference_values_for_each_item, approximate_values_for_each_item)
```
Comment (KodiaqQ) [inline nncf/quantization/algorithms/accuracy_control/ranker.py:190]: Since we use `zip` almost everywhere for `reference` and `approximate` values, can we be sure that these values match each other while every `zip`?

File: tests/openvino/tools/calibrate.py [hunk 13]
```diff
@@ -91,6 +94,12 @@ class ACValidationFunction:
     Implementation of a validation function using the Accuracy Checker.
     """
 
+    METRIC_TO_PERSAMPLE_METRIC = {
+        "coco_orig_precision": "coco_precision",
```
Comment (KodiaqQ) [inline tests/openvino/tools/calibrate.py:98]: Don't we need to match all of such kinds of metrics from POT into `calibrate.py`?
https://github.com/openvinotoolkit/openvino/blob/master/tools/pot/openvino/tools/pot/algorithms/quantization/accuracy_aware_common/utils.py#L28

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 9 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:145]: They should refer to the documentation to get more details about what validation function should return.

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 10 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:148]: Done

File: nncf/quantization/algorithms/accuracy_control/rank_functions.py (hunk 11 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/rank_functions.py:28]: Fixed

File: nncf/quantization/algorithms/accuracy_control/ranker.py (hunk 12 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/ranker.py:190]: Yes, we can. We use the same dataset to collect references values and approximate values.

File: tests/openvino/tools/calibrate.py (hunk 13 above)
Comment (andrey-churkin) [reply tests/openvino/tools/calibrate.py:98]: Done

**Comment by andrey-churkin** [comment]
run pylint pre-commit tests

**Comment by andrey-churkin** [comment]
This PR does not broke backward compatibility. A validation function that returns only single float is still supported.

**Comment by andrey-churkin** [comment]
run pytorch pre-commit tests

**Review by alexsu52 — APPROVED** [review]
